src/v/datalake/translation/scheduling.h

```diff
@@ -0,0 +1,333 @@
+ * reservation tracker to reserve resources needed for translation. Currently
+ * tracks memory but can be extended to other resources like disk usage.
+ *
+ * note: All reservations must be destroyed before destroying the tracker.
+ */
+class reservations_tracker {
+public:
+    reservations_tracker() = default;
+    reservations_tracker(const reservations_tracker&) = delete;
+    reservations_tracker& operator=(const reservations_tracker&) = delete;
+    reservations_tracker(reservations_tracker&&) = delete;
+    reservations_tracker& operator=(reservations_tracker&&) = delete;
+    virtual ~reservations_tracker() = default;
+    /**
+     * Returns a block of memory as decided by the tracker. Can be called
+     * multiple times until the requirement is satisfied.
+     */
+    virtual ss::future<reservation> reserve_memory(ss::abort_source&) = 0;
+    /* returns true if all the memory is exhausted */
+    virtual bool memory_exhausted() const = 0;
+
+    virtual size_t allocated_memory() const = 0;
+
+    static std::unique_ptr<reservations_tracker> make_default(
+      size_t total_memory, size_t memory_block_size, scheduling_notifications&);
+};
+
+/**
+ * Current status of the translator, used as input to the scheduler.
+ */
+struct translation_status {
+    clock::duration max_target_lag;
+    size_t current_offset_lag{0};
+    // Only set if a translation is currently in progress.
+    std::optional<size_t> inflight_translated_bytes;
+    std::optional<size_t> inflight_memory_bytes_reserved;
+};
+
+std::ostream& operator<<(std::ostream&, const translation_status&);
+
+/**
+ * A single schedulable translator instance. Translation interacts with
+ * two main components
+ * - a reservation tracker - to reserve resources needed for translation
+ * - a scheduler (via scheduling_notifications) - to let the scheduler know
+ *   about the translator status.
+ *
+ * The scheduler then notifies the translator when it can begin translation
+ * via \ref translator::start_translation, along with a time slice. This allows
+ * the translator to translate for the duration of the time slice.
+ *
+ * The scheduler may also preempt the inflight translation if it chooses to
+ * (via stop_translation) and the translator should stop the inflight
+ * translation and release the resources.
+ */
+class translator {
+public:
+    translator() = default;
+    translator(const translator&) = delete;
+    translator& operator=(const translator&) = delete;
+    translator(translator&&) = delete;
+    translator& operator=(translator&&) = delete;
+    virtual ~translator() = default;
+
+    /**
+     * A unique id for the translator, used as a handle for further operations.
+     */
+    virtual const translator_id& id() const = 0;
+
+    /**
+     * Invoked by the scheduler once after registration.
+     */
+    virtual ss::future<>
+    start(scheduling_notifications&, reservations_tracker&) = 0;
```

Comment:
nit: consider renaming to something like `initialize` and `finalize` to differentiate from actually `starting` an action